File: src/javascript/algorithms/strings/12.lastWordLength/lastWordLength.js
/**
 * Returns the length of the last word in a string.
 * A word is a maximal substring of non-space characters.
 * @param {string} str
 * @returns {number}
 */
function lengthOfLastWord(str) {
  let i = str.length - 1;
  let length = 0;

  // Skip trailing spaces
  while (i >= 0 && str[i] === " ") i--;

  // Count the length of the last word
  while (i >= 0 && str[i] !== " ") {
    length++;
    i--;
  }

  return length;
}

// Test cases
const testCases = [
  { str: "Welcome to DSA", expected: 3 },
  { str: " My pet is fluffy  ", expected: 6 },
  { str: "singleword", expected: 10 },
  { str: "   ", expected: 0 },
  { str: "", expected: 0 },
  { str: "a b c d", expected: 1 },
  { str: "endswithspace ", expected: 13 },
];

for (const { str, expected } of testCases) {
  const result = lengthOfLastWord(str);
  console.log(`Input: "${str}" | Output: ${result} | Expected: ${expected}`);
}

File: src/python/algorithms/strings/12.lastWordLength/last_word_length.py
"""
Length of Last Word

Given an input string str which consists of words and spaces,
return the length of the last word in the string.

Time Complexity: O(n)
Space Complexity: O(1)
"""


def length_of_last_word(s):
    """
    Traverse from end, skip trailing spaces, count last word.
    TC: O(n), SC: O(1)
    """
    i = len(s) - 1
    length = 0

    # Skip trailing spaces
    while i >= 0 and s[i] == " ":
        i -= 1

    # Count the length of the last word
    while i >= 0 and s[i] != " ":
        length += 1
        i -= 1

    return length


# Test cases
if __name__ == "__main__":
    test_cases = [
        ("Welcome to DSA", 3),
        (" My pet is fluffy  ", 6),
        ("singleword", 10),
        ("   ", 0),
        ("", 0),
        ("a b c d", 1),
        ("endswithspace ", 13),
    ]

    for s, expected in test_cases:
        result = length_of_last_word(s)
        print(f"Input: \"{s}\" | Output: {result} | Expected: {expected}")
